qutip/quantum_systems/jaynes_cummings.py

```diff
@@ -0,0 +1,168 @@
+    operators['sigma_minus'] = tensor(qeye(n_cavity), destroy(2))
+    operators['sigma_plus'] = operators['sigma_minus'].dag()
+    operators['sigma_z'] = tensor(qeye(n_cavity), sigmaz())
+    operators['sigma_x'] = tensor(qeye(n_cavity), sigmax())
+    operators['sigma_y'] = tensor(qeye(n_cavity), sigmay())
+
+    # Build Hamiltonian
+    # Free cavity evolution: ω_c a†a
+    H_cavity = omega_c * operators['n_c']
+
+    # Free atomic evolution: (ω_a/2)σ_z
+    H_atom = omega_a * operators['sigma_plus'] * operators['sigma_minus']
+
+    # Interaction term
+    if rotating_wave:
+        # Rotating wave approximation: g(a†σ_- + aσ_+)
+        H_interaction = g * (operators['a_dag'] * operators['sigma_minus'] +
+                             operators['a'] * operators['sigma_plus'])
+    else:
+        # Full interaction (Rabi model): g(a† + a)(σ_+ + σ_-)
+        H_interaction = g * (operators['a_dag'] + operators['a']) * \
+            (operators['sigma_plus'] + operators['sigma_minus'])
+
+    # Total Hamiltonian
+    hamiltonian = H_cavity + H_atom + H_interaction
+
+    # Build collapse operators for dissipation
+    c_ops = []
+
+    # Cavity decay with thermal effects 
+    if isinstance(cavity_decay, Coefficient) or (isinstance(cavity_decay, (int, float)) and cavity_decay > 0.0):
+        # Cavity relaxation: sqrt(kappa(1 + n_th)]) * a
+        if isinstance(cavity_decay, Coefficient):
+            # cavity_decay is coefficient, thermal_photons is float
+            def cavity_relax_func(t, args):
+                kappa = cavity_decay.coeff(t, args) if hasattr(cavity_decay, 'coeff') else cavity_decay(t, args)
+                return np.sqrt(kappa * (1 + thermal_photons))
+            sqrt_cavity_relax = coefficient(cavity_relax_func, args={})
+            c_ops.append(sqrt_cavity_relax * operators['a'])
+        else:
+            # Both are float
+            cavity_relax_rate = cavity_decay * (1 + thermal_photons)
+            c_ops.append(np.sqrt(cavity_relax_rate) * operators['a'])
+
+    # Cavity excitation (thermal): sqrt(kappa * n_th) * a_dag  
+    if thermal_photons > 0.0:
+        if isinstance(cavity_decay, Coefficient):
+            # cavity_decay is coefficient, thermal_photons is float
+            def cavity_excite_func(t, args):
+                kappa = cavity_decay.coeff(t, args) if hasattr(cavity_decay, 'coeff') else cavity_decay(t, args)
+                return np.sqrt(kappa * thermal_photons)
+            sqrt_cavity_excite = coefficient(cavity_excite_func, args={})
+            c_ops.append(sqrt_cavity_excite * operators['a_dag'])
+        else:
+            # Both are float
+            cavity_excite_rate = cavity_decay * thermal_photons
+            c_ops.append(np.sqrt(cavity_excite_rate) * operators['a_dag'])
+
+    # Atomic spontaneous emission: sqrt(gamma) * sigma_minus
+    if isinstance(atomic_decay, Coefficient) or (isinstance(atomic_decay, (int, float)) and atomic_decay > 0.0):
```

Comment:
Line too long. `isinstance` for detecting number is not a good idea. It could be a `np.float64`, tensorflow scalar, etc...